Sally7/RequestExecutor/ConcurrentRequestExecutor.cs

```diff
@@ -92,6 +92,14 @@ public async ValueTask<Memory<byte>> PerformRequest(ReadOnlyMemory<byte> request
                     _ = await sendSignal.WaitAsync(cancellationToken).ConfigureAwait(false);
                     try
                     {
+                        // If we bail while sending the PLC might still respond to the data that was sent.
+                        // This both breaks the send-one-receive-one flow as well as it might end up
+                        // completing a new job that reused the ID.
+#if NET5_0_OR_GREATER || NETSTANDARD2_1_OR_GREATER
+                        await
+#endif
+                        using var closeOnCancel = cancellationToken.MaybeUnsafeRegister(socket.Close);
```

Comment:
Hm, I'm not sure closing the socket is the correct approach.
In the next statement there's the send-operation on the socket. thus this could throw an `ObjectDisposedException` which is not what one would expect.

At the moment I don't have any better idea (just being back from a few days of holiday my brain isn't up on full power and I need to think the code of Sally7 a bit more...)

Reply:
It's not very nice, but I think the only clean solution would be to perform additional reads when there's no pending request for a job ID that was received. That does impact a lot of the logic in the `ConcurrentRequestExecutor` though, so I'm not sure if that's what I want to do. This implementation is not very common of course, usually there's a separate receive thread for the incoming messages and pending requests are stored in a dictionary of sorts. That's definitely an option as well...

Reply:
I'll think about an approach and share it with you (if I got some ideas).

Reply:
One approach I have in mind:
- Add the jobID to `Request`
- Add state to `Request` [free, sent, received], not sure what states are ultimately desirable
- On receive:
  - Verify Requests[&lt;received job id> % &lt;number of concurrent requests>] has matching job ID
  - Interlocked.CompareExchange request state from sent to received
    - If this fails, drop the response
  - If job ID != request.jobId, perform another response read on the socket
  - Return request to pool
- On cancel:
  - Return request to pool
    - Increment job ID with &lt;number of concurrent requests>
    - Reset state
- Replace wait on `receiveSignal` with wait on `receiveSignal` or `request.Completed`

I also think the buffer could move to `Request`, because on request complete we process the buffer sync anyway. Last but not least I think the `sendSignal` isn't required, I think we're actually allowed to invoke `socket.Send...` concurrently, because we never send multiple frames.

Reply:
This approach sounds very clean 👍🏻 

As a more or less hacky solution one could catch and swallow the `ObjectDisposedException`, as it should only ever happen once canellation happened (which is asumed to be rare).